Second layout prototype

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This is roughly what it's supposed to look like for a Group block:

![image (1)](https://github.com/WordPress/gutenberg/assets/8096000/41c83045-7795-4348-9c91-d14a623b2dc0)

Though we still need the direction control for other flex layout blocks, so for now it's staying at the top of the layout panel, with wrap next to it:

<img width="279" alt="Screenshot 2023-09-04 at 1 16 50 pm" src="https://github.com/WordPress/gutenberg/assets/8096000/effe1c54-a10b-49bd-af78-8cd11f25f36f">

The segmented controls are still the ones we have currently, so don't look exactly like the design.

Padding is still in the dimensions panel, but for now the focus should be on testing the layout controls and then we can work out if/how to move the other dimensions controls into the layout panel.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Test all blocks with layout and check that controls work and their positioning makes sense.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Remove variation controls and allow switching on group

tellthemachines committed:
Add some controls

tellthemachines committed:
Move dimensions panel

tellthemachines committed:
Fix controls

tellthemachines committed:
Width and height controls not working yet

tellthemachines committed:
Functionality is half there

tellthemachines committed:
prevent error

tellthemachines committed:
Width/height controls working

tellthemachines committed:
It more or less works!!

tellthemachines committed:
Fix matrix resetting values

tellthemachines committed:
Fix justify value

tellthemachines committed:
reset justify and align on orientation change

tellthemachines committed:
Fix defaults

tellthemachines committed:
self width for stack children

tellthemachines committed:
Align Row children

tellthemachines committed:
Make proper fixed sizes

tellthemachines committed:
constrained layout child controls logic

SaxonF committed:
persistent width and height

SaxonF committed:
alignment options layout

tellthemachines committed:
Add back group variation controls

tellthemachines committed:
Vertical and horizontal instead of matrix

tellthemachines committed:
Remove comment

tellthemachines committed:
Add icons to vertical and horizontal

tellthemachines committed:
Update gap control

tellthemachines committed:
make icons pretty

tellthemachines committed:
Conditionally display horizontal/vertical

tellthemachines committed:
Fix error in blocks with full align support

andrewserong reviewed (commented):
Very cool exploration so far! I like how it reduces the differences in the UI between each of the layout types. It does change quite a lot of the UI, and moving Dimensions to a different tab could be a fairly big change for folks, so very curious to hear what folks think about the design direction, but overall it's looking like a lot of promising work to me:

| Group | Row |
| --- | --- |
| <img src="https://github.com/WordPress/gutenberg/assets/14988353/8e5a5cff-9701-449d-8aee-205af832db5b" width="320" /> | <img src="https://github.com/WordPress/gutenberg/assets/14988353/5aa34561-1f3d-4e20-98b8-f2ad6632e257" width="320" /> |

I haven't had a close look at the code since it's still in progress, but I had a couple of quick questions from a first play through:

* Is the plan still to switch the Layout panel over to use the ToolsPanel? The designs have the ellipsis at the top right of the Layout panel, so I wasn't sure if the goal was to do that at this stage.
* Will it still be possible for folks to set content + wide size as before? I could see where to add width: fixed under Dimensions, but wasn't sure how this would work with existing Group or other blocks that have custom content/wide size that differs from an overall theme's values.
* What does "boxed" mean under Content Width?
* You can now set Fixed height and Min Height at the same time for the Group block. What's the difference between these?

I think probably the main concerns I'd have personally about the changes are about how it's going to feel for folks used to the existing controls transitioning to the new ones and the impact of the paradigm shift in how layout works. I imagine shifting the paradigm somewhat is the goal here, so some friction is likely to be expected, but still curious about how it'll feel for folks reaching for these tools to build out some designs. Great work, though, keen to see how this progresses!

tellthemachines commented:
Thanks for the feedback @andrewserong !

> Is the plan still to switch the Layout panel over to use the ToolsPanel? The designs have the ellipsis at the top right of the Layout panel, so I wasn't sure if the goal was to do that at this stage.

Yes I think so, though I'm not 100% sure all those options will be resettable. 

> Will it still be possible for folks to set content + wide size as before? I could see where to add width: fixed under Dimensions, but wasn't sure how this would work with existing Group or other blocks that have custom content/wide size that differs from an overall theme's values.

Yes! @SaxonF is finalising the design for the custom width inputs. I think we mostly agree that they shouldn't be as prominent as they currently are, but not sure what they'll end up looking like.

> What does "boxed" mean under Content Width?

It means the there is a max content width (as opposed to content taking up full width as in flow layout), whether it's the theme one or a custom one.

> You can now set Fixed height and Min Height at the same time for the Group block. What's the difference between these?

Ah yes! Fixed height should use `height` or `flex-basis` with `flex-shrink: 0` so that the height is effectively fixed, whereas `min-height` is the existing control that sets a minimum height that is expandable depending on contents. Fixed height can be useful on Stacks to enable certain layout types or if wanting to force a scrollbar I guess 😄

SaxonF committed:
update content and wrap controls

SaxonF committed:
fix alignment and spacing layout controls

SaxonF committed:
fix width/height bug

tellthemachines committed:
Fix error when block supports align is false

tellthemachines committed:
Fix wrap bug

tellthemachines committed:
Fix rebase issues

tellthemachines committed:
Fix align

tellthemachines committed:
Skip pesky tests

tellthemachines committed:
Fix e2e test

tellthemachines committed:
Fix alignment

tellthemachines committed:
Remove redundant props

tellthemachines committed:
Remove flex-basis

SaxonF committed:
fill width updates

tellthemachines commented:
This is quite a big changeset, and it's hard to keep the PR in sync with a continually changing codebase, so I'm thinking of splitting this out into several pieces. We can reasonably easily split out the following:

* Adding Icon support to `CustomSelectControl` (and/or its experimental V2)

* Adding Width and Height to ChildLayoutControl 
     * as replacement for Fit/Filled/Fixed in flex layouts 
     * as a new feature in constrained/flow layouts
     * probably not in Grid layouts, because column and row span have a completely different logic?
     * decide whether to change `allowSizingOnChildren` to opt-out instead of opt-in

* Moving over at least the layout-related controls from Dimensions to Layout panels
    * do we move also Padding, Margin etc. and unify both panels?
    * should they live in Styles or Settings?

* Updates to the existing controls currently in the Layout panel
    * Direction becomes effectively a layout switcher (still configurable via `allowSwitching`)
    * Content width control replaces "Inner blocks use content width"
    * Horizontal and Vertical selects replace justification/alignment
    * Wrap control gets new look.
